File: src/main/java/com/micabytes/ink/Stitch.java
package com.micabytes.ink;

import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;

class Stitch extends ParameterizedContainer {

  @SuppressWarnings("StringBufferReplaceableByString")
  Stitch(int l, String str, @Nullable Container current) throws InkParseException {
    lineNumber = l;
    level = 1;
    if (current == null)
      throw new InkParseException("A stitch cannot be defined without a parent Knot");
    parent = current.getContainer(0);
    parent.add(this);
    type = ContentType.STITCH;
    String fullId = new StringBuilder(parent.id)
        .append(InkParser.DOT)
        .append(str.replaceAll(String.valueOf(InkParser.HEADER), "").trim())
        .toString();
    if (fullId.contains(StoryText.BRACE_LEFT)) {
      String params = fullId.substring(fullId.indexOf(StoryText.BRACE_LEFT)+1, fullId.indexOf(StoryText.BRACE_RIGHT));
      String[] param = params.split(",");
      parameters = new ArrayList<>();
      for (String aParam : param) getParameters().add(aParam.trim());
      fullId = fullId.substring(0, fullId.indexOf(StoryText.BRACE_LEFT));
    }
    //if (fullId.startsWith(FUNCTION)) {
    //
    //}
    id = fullId;
  }

  public static boolean isStitchHeader(String str) {
    if (str.length() < 2) return false;
    return str.charAt(0) == InkParser.HEADER && str.charAt(1) != InkParser.HEADER;
  }

}

File: src/main/kotlin/com/micabytes/ink/Stitch.kt
package com.micabytes.ink

internal class Stitch(header: String,
                      parent: Container,
                      lineNumber: Int) : ParameterizedContainer(getId(header, parent), ParameterizedContainer.getParameters(header), parent, lineNumber) {

  companion object {
    private val STITCH_HEADER = "="

    fun isStitchHeader(str: String): Boolean {
      return str.startsWith(STITCH_HEADER)
    }

    fun  getId(header: String, parent: Container): String {
      var id = header.replace(STITCH_HEADER.toRegex(), "").trim({ it <= ' ' })
      if (id.startsWith(Symbol.FUNCTION)) {
        id = id.replaceFirst(Symbol.FUNCTION.toRegex(), "")
      }
      if (id.contains(Symbol.BRACE_LEFT)) {
        id = id.substring(0, id.indexOf(Symbol.BRACE_LEFT))
      }
      return parent.id + Symbol.DOT + id
    }

    fun getParent(currentContainer: Container): Container {
      var container = currentContainer
      while (container != null) {
        if (container is Knot) return container
        container = container.parent!!
      }
      return currentContainer
    }

  }

}
